src/framework/components/element/component.js

```diff
@@ -86,50 +168,109 @@ pc.extend(pc, function () {
+
+            this.element._width = rect.z - rect.x;
+            this.element._height = rect.w - rect.y;
+
+            // the rect is going to be Vec4 storing the following values:
+            // [ left offset, bottom offset, right offset, top offset ]
+            if (this.element._anchorDirty || this.element._cornerDirty) {               
+                this.element._anchorTransform.setTranslate(rect.x, rect.y, 0);
                 this.element._anchorDirty = false;
             }
 
             if (this.dirtyWorld) {
+                // before recomputing the transforms let's agree on a few matrices used below:
+                //
+                //    * world: it's either clip box of the WebGL (for screen and camera screen types) OR
+                //             real world box (for world screen type).
+                //             basically, the "ouput" coords sans local transforms of the element
+                //    * _screenToWorld: transforms screen point to the "world" point
+                //    * _modelTransform: transforms screen point further down the heirarchy
+                //    * localTransform: this is normal entity transforms (local, of course)
+                //    * _anchorTransform: just the offset to satisfy anchoring settings (the offset of lower left corner)
+                //    * toPivotTransform: just the offset to pivot point of the element
+                //
                 if (this._parent === null) {
-                    this.worldTransform.copy(this.localTransform);
+                    // no parent? _screenToWorld is basically the local transform
+                    this.element._screenToWorld.copy(this.localTransform);
                 } else {
-                    // transform element hierarchy
+                    // ok, we have a parent. does it own an element?
+                    // TODO: lookup up to the scene root would be more correct – what if there is a blank 
+                    //       object between two elements?
                     if (this._parent.element) {
+                        // our _screenToWorld starts off by offsetting current transform (which is parent's) by
+                        // anchor offset – like we move the box to match the anchor settings first
                         this.element._screenToWorld.mul2(this._parent.element._modelTransform, this.element._anchorTransform);
                     } else {
+                        // no element means we start with plain anchoring transform
                         this.element._screenToWorld.copy(this.element._anchorTransform);
                     }
 
-                    this.element._modelTransform.mul2(this.element._screenToWorld, this.localTransform);
+                    // let's compute the pivot point – remember it's local to element coord space
+                    var pivotPoint          = new pc.Vec3( this.element._width * this.element.pivot.x, this.element._height * this.element.pivot.y, 0 );
```

Comment:
I will check what we can pre-alloc – I guess pretty much everything. Thank you for the hint